tensorflow_io/core/kernels/http_kernels.cc

```diff
@@ -13,16 +13,527 @@ See the License for the specific language governing permissions and
+  };
+  DirectResponseState direct_response_ = {};
+
+  CURL* curl_ = nullptr;
+  curl_slist* curl_headers_ = nullptr;
+  curl_slist* resolve_list_ = nullptr;
+
+  std::unordered_map<std::string, std::string> response_headers_;
+  uint64_t response_code_ = 0;
+
+  std::string uri_;
+
+  // The timestamp of the last activity related to the request execution, in
+  // seconds since epoch.
+  uint64_t last_progress_timestamp_ = 0;
+  // The last progress in terms of bytes transmitted.
+  curl_off_t last_progress_bytes_ = 0;
+
+  // The maximum period of request inactivity.
+  uint32_t inactivity_timeout_secs_ = 60;  // 1 minute
+
+  // Timeout for the connection phase.
+  uint32_t connect_timeout_secs_ = 120;  // 2 minutes
+
+  // Timeout for the whole request. Set only to prevent hanging indefinitely.
+  uint32_t request_timeout_secs_ = 3600;  // 1 hour
+
+  // Limit the size of an http response that is copied into an error message.
+  const size_t response_to_error_limit_ = 500;
+
+  bool IsDirectResponse() const { return direct_response_.buffer_ != nullptr; }
+
+  absl::string_view GetResponse() const {
+    absl::string_view response;
+    if (IsDirectResponse()) {
+      response = absl::string_view(direct_response_.buffer_,
+                                   direct_response_.bytes_transferred_);
+    } else {
+      response =
+          absl::string_view(response_buffer_.data(), response_buffer_.size());
+    }
+    return response;
+  }
+
+  // Cancels the transmission if no progress has been made for too long.
+  static int ProgressCallback(void* this_object, curl_off_t dltotal,
+                              curl_off_t dlnow, curl_off_t ultotal,
+                              curl_off_t ulnow) {
+    auto that = reinterpret_cast<CurlHttpRequest*>(this_object);
+    const int64_t now = absl::ToUnixSeconds(absl::Now());
+    const auto current_progress = dlnow + ulnow;
+    if (that->last_progress_timestamp_ == 0 ||
+        current_progress > that->last_progress_bytes_) {
+      // This is the first time the callback is called or some progress
+      // was made since the last tick.
+      that->last_progress_timestamp_ = now;
+      that->last_progress_bytes_ = current_progress;
+      return 0;
+    }
+
+    if (now - that->last_progress_timestamp_ > that->inactivity_timeout_secs_) {
+      double lookup_time = -1;
+      const auto lookup_time_status = curl_easy_getinfo(
+          that->curl_, CURLINFO_NAMELOOKUP_TIME, &lookup_time);
+
+      double connect_time = -1;
+      const auto connect_time_status =
+          curl_easy_getinfo(that->curl_, CURLINFO_CONNECT_TIME, &connect_time);
+
+      double pretransfer_time = -1;
+      const auto pretransfer_time_status = curl_easy_getinfo(
+          that->curl_, CURLINFO_PRETRANSFER_TIME, &pretransfer_time);
+
+      double starttransfer_time = -1;
+      const auto starttransfer_time_status = curl_easy_getinfo(
+          that->curl_, CURLINFO_PRETRANSFER_TIME, &starttransfer_time);
```

Comment:
should be `CURLINFO_STARTTRANSFER_TIME` ?

Reply:
Thanks. Updated.